backend/json_test.go

```diff
@@ -43,14 +45,90 @@ func TestResponseEncoder(t *testing.T) {
 	require.Equal(t, str, string(b2), "same result from pointer or object")
 
 	// Now the same thing in query data
-	qdr := NewQueryDataResponse()
+	qdr := backend.NewQueryDataResponse()
 	qdr.Responses["A"] = dr
 
 	b, err = json.Marshal(qdr)
 	require.NoError(t, err)
 
 	str = string(b)
-	require.Equal(t, `{"responses":{"A":{"frames":[{"schema":{"name":"simple","fields":[{"name":"time","type":"time","typeInfo":{"frame":"time.Time"}},{"name":"valid","type":"bool","typeInfo":{"frame":"bool"}}]},"data":{"values":[[1577934240000,1577934300000],[true,false]]}},{"schema":{"name":"other","fields":[{"name":"value","type":"number","typeInfo":{"frame":"float64"}}]},"data":{"values":[[1]]}}]}}}`, str)
+	require.Equal(t, `{"results":[{"refId":"A","frames":[{"schema":{"name":"simple","fields":[{"name":"time","type":"time","typeInfo":{"frame":"time.Time"}},{"name":"valid","type":"bool","typeInfo":{"frame":"bool"}}]},"data":{"values":[[1577934240000,1577934300000],[true,false]]}},{"schema":{"name":"other","fields":[{"name":"value","type":"number","typeInfo":{"frame":"float64"}}]},"data":{"values":[[1]]}}]}]}`, str)
+
+	// Read the parsed result and make sure it is the same
+	copy := &backend.QueryDataResponse{}
+	err = json.Unmarshal(b, copy)
+	require.NoError(t, err)
+	require.Equal(t, len(qdr.Responses), len(copy.Responses))
+
+	// Check the final result
+	for k, val := range qdr.Responses {
+		other := copy.Responses[k]
+		require.Equal(t, len(val.Frames), len(other.Frames))
+
+		for idx := range val.Frames {
+			a := val.Frames[idx]
+			b := other.Frames[idx]
+
+			if diff := cmp.Diff(a, b, data.FrameTestCompareOptions()...); diff != "" {
+				t.Errorf("Result mismatch (-want +got):\n%s", diff)
+			}
+		}
+	}
+}
+
+func checkResultOrder(t *testing.T, qdr *backend.QueryDataResults, expectedOrder ...string) {
+	t.Helper()
+
+	raw, err := json.Marshal(qdr)
+	require.NoError(t, err)
+
+	dummy := make(map[string]interface{})
+	err = json.Unmarshal(raw, &dummy)
+	require.NoError(t, err)
+
+	arr, ok := dummy["results"].([]interface{})
+	require.True(t, ok)
+
+	require.Equal(t, len(expectedOrder), len(arr))
+
+	found := make([]string, 0)
+	for idx := range arr {
+		res, ok := arr[idx].(map[string]interface{})
+		require.True(t, ok)
+
+		key, ok := res["refId"].(string)
+		require.True(t, ok)
+
+		found = append(found, key)
+	}
+
+	require.EqualValues(t, expectedOrder, found)
+}
+
+// TestResponseEncoder makes sure that the JSON produced from arrow and dataframes match
+func TestResponseEncoderForOrder(t *testing.T) {
+	rsp := backend.NewQueryDataResponse()
+	rsp.Responses["A"] = backend.DataResponse{}
+	rsp.Responses["B"] = backend.DataResponse{}
+
+	qdr := &backend.QueryDataResults{
+		Results: rsp.Responses,
+	}
+	checkResultOrder(t, qdr, "A", "B")
```

Comment:
you can't rely on this order since there is an iteration over map internally, Go randomizes iteration over a map.